### PR TITLE
[FIX] use sklearn html representation of estimators

### LINKED ISSUE #5924: [BUG] HTML representation of estimator parameters do not show in notebooks
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating system

- [x] Linux
- [ ] Mac
- [ ] Windows

### Operating system version

For example one of the following:
- Linux Ubuntu 22.04

### Python version

- [ ] 3.14
- [x] 3.13
- [ ] 3.12
- [ ] 3.11
- [ ] 3.10

### nilearn version

dev

### Expected behavior

since sklearn 1.8 estimators can have an improved HTML representation

https://scikit-learn.org/stable/auto_examples/release_highlights/plot_release_highlights_1_8_0.html#html-representation-of-estimators

I would expect that the Nilearn estimators should be able to take advantage of this but it does not see to be the case

### Current behavior & error messages

This is what I got (not visible on the image but the parameters show up their documentation when hovering over them).

Image

### Steps and code to reproduce bug

tried with this marimo notebook

```python
import marimo

__generated_with = "0.18.4"
app = marimo.App(width="full")


@app.cell
def _():
    import numpy as np
    import nibabel as nib
    from nilearn.decoding import Decoder

    rng = np.random.RandomState(42)
    X = rng.rand(5, 5, 5, 20)
    X_img = nib.Nifti1Image(X, np.eye(4))
    y = np.array([0] * 10 + [1] * 10)

    decoder = Decoder()
    decoder.fit(X_img, y)
    return


@app.cell
def _():
    from sklearn import linear_model

    reg = linear_model.LinearRegression()

    reg.fit([[0, 0], [1, 1], [2, 2]], [0, 1, 2])
    return


if __name__ == "__main__":
    app.run()
```

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #5924

Changes proposed in this pull request:

- [x] add a NilearnBaseEstimator
- [x] make sure that the URL to the doc of the each estimator doc can be generated
- [x] rely on the sklearn HTML representation of estimator in reports (for sklearn >= 1.7.0)
- [x] show pictures of before and after
- [x] changelog

### CODE REVIEW

**Comment by Remi-Gau** [comment]
There seems to be a small upstream issue: https://github.com/scikit-learn/scikit-learn/issues/33005

File: nilearn/reporting/html_report.py [hunk 1]
```diff
@@ -264,14 +261,6 @@ def _create_report(
...
                 sparsify=False,
             )
-    parameters = model_attributes_to_dataframe(estimator)
-    with pd.option_context("display.max_colwidth", 100):
-        parameters = dataframe_to_html(
-            parameters,
-            precision=2,
-            header=True,
-            sparsify=False,
-        )
```
Comment (Remi-Gau) [inline nilearn/reporting/html_report.py:274]: we are going to need this for older version of sklearn that do not have an html repr of the estimators parameter

File: nilearn/_utils/html_repr.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            version_url = (
+                f"{nil_version.major}.{nil_version.minor}.{nil_version.minor}"
+            )
+
+        return (
+            f"https://nilearn.github.io/{version_url}/modules/generated/"
+            "{estimator_module}.{estimator_name}.html"
+        )
+    
+    if SKLEARN_GTE_1_8:
```
Comment (Remi-Gau) [inline nilearn/_utils/html_repr.py]: this may work with some older version of sklearn: find cut off

File: nilearn/glm/_base.py [hunk 3]
```diff
@@ -48,13 +49,34 @@
...
 
-class BaseGLM(CacheMixin, BaseEstimator):
+class BaseGLM(_NilearnHTMLDocumentationLinkMixin, CacheMixin, BaseEstimator):
     """Implement a base class \
     for the :term:`General Linear Model<GLM>`.
     """
 
     _estimator_type = "glm"  # TODO (sklearn >= 1.8) remove
 
+    if SKLEARN_GTE_1_8:
```
Comment (Remi-Gau) [inline nilearn/glm/_base.py]: this may work with some older version of sklearn: find cut off

File: nilearn/glm/_base.py [hunk 4]
```diff
@@ -537,15 +559,18 @@ def generate_report(
...
-
-        model_attributes = glm_model_attributes_to_dataframe(self)
-        with pd.option_context("display.max_colwidth", 100):
-            model_attributes_html = dataframe_to_html(
-                model_attributes,
-                precision=2,
-                header=True,
-                sparsify=False,
-            )
+        if SKLEARN_GTE_1_8:
```
Comment (Remi-Gau) [inline nilearn/glm/_base.py]: this may work with some older version of sklearn: find cut off

File: nilearn/maskers/tests/test_html_report.py [hunk 5]
```diff
@@ -74,7 +74,8 @@ def generate_and_check_masker_report(
         excludes.append('<div class="image">')
 
     else:
-        if masker.__sklearn_is_fitted__():
+        if not SKLEARN_GTE_1_8 and masker.__sklearn_is_fitted__():
```
Comment (Remi-Gau) [inline nilearn/maskers/tests/test_html_report.py]: this may work with some older version of sklearn: find cut off

also make sure there is a else block here to check content when relying on the sklearn html repr

**Comment by Remi-Gau** [comment]
actually commit [db92f29](https://github.com/nilearn/nilearn/pull/5925/commits/db92f2937e1bf61fb12987ad3f6e470b980f37d3) may not be necessary given it is a bad hack to fix a 1.8.0 bug

**Comment by Remi-Gau** [comment]
pictures to compare

before: 
image


after:

image

We are losing the readibility of some parameters, but we are gaining (on top of less code to maintain for us) the docstring of each parameter when hovering over each param and the hightlighting of parameters that do not have the default value.

Note that some of the funky looking things about t_r in the new table is a bug in sklearn 1.8.0 that should be fixed in 1.8.1.

File: nilearn/_utils/estimator_checks.py [hunk 6]
```diff
@@ -989,6 +991,35 @@ def check_doc_attributes(estimator) -> None:
         )
 
 
+def check_doc_link(estimator_orig):
```
Comment (bthirion) [inline nilearn/_utils/estimator_checks.py:1010]: in that case, it could be useful to describe the type of estimator_orig in the docstring.

File: nilearn/_utils/html_repr.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    _doc_link_module = "nilearn"
+
+    @property
+    def _doc_link_template(self):
+        version_url = "dev"
+        nil_version = parse(__version__)
+        if nil_version.dev is None:
+            version_url = (
+                f"{nil_version.major}.{nil_version.minor}.{nil_version.minor}"
```
Comment (bthirion) [inline nilearn/_utils/html_repr.py]:
```suggestion
                f"{nil_version.major}.{nil_version.minor}.{nil_version.patch}"
```

Comment (Remi-Gau) [reply nilearn/_utils/html_repr.py]: note that the versiion object does not have patch attribute: https://packaging.pypa.io/en/stable/version.html#packaging.version.Version

**Comment by Remi-Gau** [comment]
> I could not see the output in the built doc.

@bthirion 

see for example the reports for the maskers:

https://output.circle-artifacts.com/output/job/7b0be4b3-0b4e-4b69-a624-cd60df63fdc8/artifacts/0/dev/modules/generated_reports/masker_reports_examples.html

or 

https://output.circle-artifacts.com/output/job/7b0be4b3-0b4e-4b69-a624-cd60df63fdc8/artifacts/0/dev/_static/notebook_reports_maskers.html

**Comment by Remi-Gau** [comment]
> > I could not see the output in the built doc.
> 
> @bthirion
> 
> see for example the reports for the maskers:
> 
> https://output.circle-artifacts.com/output/job/7b0be4b3-0b4e-4b69-a624-cd60df63fdc8/artifacts/0/dev/modules/generated_reports/masker_reports_examples.html
> 
> or
> 
> https://output.circle-artifacts.com/output/job/7b0be4b3-0b4e-4b69-a624-cd60df63fdc8/artifacts/0/dev/_static/notebook_reports_maskers.html

@emdupre 
what do you think of the using the sklearn HTML representation of estimators in our reports (or for notebooks)?

Not visible in the doc here, but with sklearn >=1.8.0 users would also get the docstring of each attribute when hovering over the table.

**Comment by emdupre** [comment]
> @emdupre
what do you think of the using the sklearn HTML representation of estimators in our reports (or for notebooks)?

> We are losing the readibility of some parameters, but we are gaining (on top of less code to maintain for us) the docstring of each parameter when hovering over each param and the hightlighting of parameters that do not have the default value.

I think this is a reasonable tradeoff, particularly given the decrease in maintenance cost ! Personally, it's a bit hard for me to intuit why some values are highlighted without the roll-over feature, but I don't think that's worth bumping our sklearn version for.

Should we note in the documentation somewhere that non-default values are highlighted, since we can't assume rollover works ?

**Comment by Remi-Gau** [comment]
> Should we note in the documentation somewhere that non-default values are highlighted, since we can't assume rollover works ?

fine by me but where in the doc given how many estimators are affected

**Comment by Remi-Gau** [comment]
> but I don't think that's worth bumping our sklearn version for.

agreed that this does not justify it: but the number of TODOs that are dependent on a sklearn is slowly growing (at least 50 on main already) : SPEC-0 suggests dropping support for sklearn in the first quarter of 2026 (https://scientific-python.org/specs/spec-0000/), so maybe we could drop it at the end of 2026... to slowly move forward...

**Comment by bthirion** [comment]
> > but I don't think that's worth bumping our sklearn version for.
> 
> agreed that this does not justify it: but the number of TODOs that are dependent on a sklearn is slowly growing (at least 50 on main already) : SPEC-0 suggests dropping support for sklearn in the first quarter of 2026 (https://scientific-python.org/specs/spec-0000/), so maybe we could drop it at the end of 2026... to slowly move forward...

you probably mean sklearn-1.4

**Comment by Remi-Gau** [comment]
> you probably mean sklearn-1.4

indeed !!!! 🤣

**Review by hndgzkn — APPROVED** [review]
LGTM. 

I am not sure if that's a good idea but might worth discussion. 

How does it sound to create a `BaseNilearnEstimator` and inherit from `_NilearnHTMLDocumentationLinkMixin`? That would avoid updating each estimator and also help us implement common code for all estimators.

**Comment by Remi-Gau** [comment]
> How does it sound to create a `BaseNilearnEstimator` and inherit from `_NilearnHTMLDocumentationLinkMixin`? That would avoid updating each estimator and also help us implement common code for all estimators.

not sure if it would be very different: at the moment we can just update _NilearnHTMLDocumentationLinkMixin and all classes that use it will be affected, no?

**Comment by hndgzkn** [comment]
> not sure if it would be very different: at the moment we can just update _NilearnHTMLDocumentationLinkMixin and all classes that use it will be affected, no?

If a change is required on _NilearnHTMLDocumentationLinkMixin functionality, yes, this fixes it. 

However in the future, if we need a new mixin to be extended by all estimators, we will need to put it in all estimators and create a new dependency for each.
 
Or, if there will be a behavior different than BaseEstimator of scikit-learn, we will need to put it in all estimator classes.  

For example, if we had a base class for nilearn estimators, we could define `_more_tags` and common parts of `__sklearn_tags__`there and we would not need to implement it in all estimators. 

In this PR, we are introducing a new dependency in all estimators. However, if we are going to make a change in all estimators, this PR would be an opportunity to remove direct dependency on sklearn `BaseEstimator` and make each estimator dependent only on `BaseNilearnEstimator` class. It would be a central place to make changes common to all estimators and avoid changing all estimators each time.

**Comment by Remi-Gau** [comment]
> we could define `_more_tags` and common parts of `__sklearn_tags__`

hm... let me try with _more_tags first

less convinced about __sklearn_tags__

**Comment by hndgzkn** [comment]
> less convinced about **sklearn_tags**

Yes, might not be possible, or simpler.

The main benefit would be to reduce dependencies for estimator classes and avoid touching them again when there is a change that would apply to all.

File: nilearn/_base.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+
+class NilearnBaseEstimator(_NilearnHTMLDocumentationLinkMixin, BaseEstimator):
+    def _more_tags(self):
+        """Return estimator tags.
+
+        TODO (sklearn >= 1.6.0) remove
+        """
+        return self.__sklearn_tags__()
```
Comment (Remi-Gau) [inline nilearn/_base.py]: added here and removed from all our estimators

**Comment by Remi-Gau** [comment]
@emdupre @bthirion 
asking for another look since this PR now adds a `NilearnBaseEstimator` that all our estimator inherit from.

**Comment by emdupre** [comment]
> fine by me but where in the doc given how many estimators are affected

Maybe we should have an example on the reports in our doc (thinking in [Viz](https://nilearn.github.io/stable/auto_examples/01_plotting/index.html)) similarly to the sklearn section on [Displaying Pipelines](https://scikit-learn.org/stable/auto_examples/miscellaneous/plot_pipeline_display.html). It would really just be giving visibility to the [nilearn.reporting examples](https://nilearn.github.io/stable/modules/reporting.html).

Not for this PR, but we can open a new issue if you agree !

**Comment by emdupre** [comment]
> @emdupre @bthirion
asking for another look since this PR now adds a NilearnBaseEstimator that all our estimator inherit from.

I think the `NilearnBaseEstimator` makes sense ; agree that it will just be good to get the tests passing to confirm !

**Comment by Remi-Gau** [comment]
> Maybe we should have an example on the reports in our doc (thinking in [Viz](https://nilearn.github.io/stable/auto_examples/01_plotting/index.html)) similarly to the sklearn section on [Displaying Pipelines](https://scikit-learn.org/stable/auto_examples/miscellaneous/plot_pipeline_display.html). It would really just be giving visibility to the [nilearn.reporting examples](https://nilearn.github.io/stable/modules/reporting.html).
> 
> Not for this PR, but we can open a new issue if you agree !

actually I remembered seeing the sklearn html repr of some our estimators in our examples:

https://nilearn.github.io/dev/auto_examples/04_glm_first_level/plot_bids_features.html#first-level-model-estimation-one-subject

image


so in a follow up PR we could make sure the very simplest (entry level) examples with maskers, decoders, glm... actually display this HTML representation of the estimator with some explanation of the color coding

**Comment by Remi-Gau** [comment]
OK this is an example with decoder:

https://output.circle-artifacts.com/output/job/5eed32a0-d37c-4a96-b038-615be4704c1b/artifacts/0/dev/auto_examples/00_tutorials/plot_decoding_tutorial.html#decoding-with-support-vector-machine

image

Will do the same for one GLM and one masker

**Comment by Remi-Gau** [comment]
@emdupre 
which example would you consider the best 'entry level' one that explains 'maskers' to users?
we have something for decoding and glm but nothing for maskers in `examples/00_tutorials`

**Comment by emdupre** [comment]
Usually I point to this section of the User Guide : https://nilearn.github.io/stable/manipulating_images/masker_objects.html

There we point to these examples : 
1. https://nilearn.github.io/stable/auto_examples/06_manipulating_images/plot_mask_computation.html#sphx-glr-auto-examples-06-manipulating-images-plot-mask-computation-py
2. https://nilearn.github.io/stable/auto_examples/06_manipulating_images/plot_nifti_simple.html#sphx-glr-auto-examples-06-manipulating-images-plot-nifti-simple-py

So I'd suggest one of those, personally ! Given the way we structured the docs so far, I don't know that I would put anything in 00_tutorials

**Comment by Remi-Gau** [comment]
> Usually I point to this section of the User Guide : https://nilearn.github.io/stable/manipulating_images/masker_objects.html
> 
> There we point to these examples :
> 
>     1. https://nilearn.github.io/stable/auto_examples/06_manipulating_images/plot_mask_computation.html#sphx-glr-auto-examples-06-manipulating-images-plot-mask-computation-py
> 
>     2. https://nilearn.github.io/stable/auto_examples/06_manipulating_images/plot_nifti_simple.html#sphx-glr-auto-examples-06-manipulating-images-plot-nifti-simple-py
> 
> 
> So I'd suggest one of those, personally ! Given the way we structured the docs so far, I don't know that I would put anything in 00_tutorials

Thanks @emdupre that's very useful.
I will rework those examples in this PR.

I actually feel that the [plot-nifti-simple-py](https://nilearn.github.io/stable/auto_examples/06_manipulating_images/plot_nifti_simple.html#sphx-glr-auto-examples-06-manipulating-images-plot-nifti-simple-py) could be moved to the tutorial section.

**Comment by Remi-Gau** [comment]
> I actually feel that the [plot-nifti-simple-py](https://nilearn.github.io/stable/auto_examples/06_manipulating_images/plot_nifti_simple.html#sphx-glr-auto-examples-06-manipulating-images-plot-nifti-simple-py) could be moved to the tutorial section.

maybe in another PR

File: doc/maintenance.rst [hunk 9]
```diff
@@ -279,6 +280,7 @@ but care should be taken that at least the following have not been affected:
 
 - ``.rst`` files containing the string ``image:: ../auto_examples``,
 - the tables in the :ref:`plotting` section,
+- ``literalinclude`` sphinx-directives with a ``:start-after:``
```
Comment (Remi-Gau) [inline doc/maintenance.rst:283]: adding this for us to check before making a release

File: doc/manipulating_images/masker_objects.rst [hunk 10]
```diff
@@ -148,10 +148,6 @@ method of the masker. The generated report can be viewed in a Jupyter notebook,
 opened in a new browser tab using ``report.open_in_browser()``,
 or saved as a portable HTML file ``report.save_as_html(output_filepath)``.
 
-.. literalinclude:: ../../examples/06_manipulating_images/plot_mask_computation.py
-    :start-after: # We need to specify an 'epi' mask_strategy, as this is raw :term:`EPI` data
-    :end-before: # Generate mask with strong opening
-
```
Comment (Remi-Gau) [inline doc/manipulating_images/masker_objects.rst:154]: removing this literal include because in the end it just amounted to showing

```python
masker.generate_report()
```

File: doc/user_guide.rst [hunk 11]
```diff
@@ -19,7 +19,7 @@ Table of contents
 
 .. toctree::
    :numbered:
-   :maxdepth: 3
+   :maxdepth: 2
```
Comment (Remi-Gau) [inline doc/user_guide.rst:22]: just to avoid a TOC with some many entries that it requires doom scrolling through 25 pages

https://nilearn.github.io/stable/user_guide.html

**Comment by Remi-Gau** [comment]
@emdupre I cleaned a bit the 2 examples you mentioned to also add some more headings

**Comment by emdupre** [comment]
>> I actually feel that the [plot-nifti-simple-py](https://nilearn.github.io/stable/auto_examples/06_manipulating_images/plot_nifti_simple.html#sphx-glr-auto-examples-06-manipulating-images-plot-nifti-simple-py) could be moved to the tutorial section.

> maybe in another PR

@Remi-Gau yes let's discuss in another issue ; to me it could be related to https://github.com/nilearn/nilearn/issues/3379
